src/libopensc/pkcs15-esteid2018.c

```diff
@@ -0,0 +1,220 @@
+	if (*strp)
+		free(*strp);
+	*strp = value ? strdup(value) : NULL;
+}
+
+static int sc_pkcs15emu_esteid2018_init(sc_pkcs15_card_t *p15card) {
+	sc_card_t *card = p15card->card;
+	u8 buff[128];
+	int r, i;
+	size_t field_length = 0, taglen, j;
+	sc_path_t tmppath;
+
+	set_string(&p15card->tokeninfo->label, "ID-kaart");
+	set_string(&p15card->tokeninfo->manufacturer_id, "IDEMIA");
+
+	/* Read docnr */
+	sc_format_path("3F00D003", &tmppath);
+	LOG_TEST_RET(card->ctx, sc_select_file(card, &tmppath, NULL), "SELECT docnr");
+	r = sc_read_binary(card, 0, buff, 11, 0);
+	LOG_TEST_RET(card->ctx, r, "read docnr failed");
+	const unsigned char *tag = sc_asn1_find_tag(card->ctx, buff, (size_t)r, 0x04, &taglen);
+	if (tag == NULL)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	for (j = 0; j < taglen; j++)
+		if (!isalnum(tag[j]))
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	p15card->tokeninfo->serial_number = malloc(taglen + 1);
+	if (!p15card->tokeninfo->serial_number)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+	p15card->tokeninfo->serial_number = memcpy(p15card->tokeninfo->serial_number, tag, taglen);
+	p15card->tokeninfo->serial_number[taglen] = '\0';
+	p15card->tokeninfo->flags = SC_PKCS15_TOKEN_READONLY;
+
+	/* add certificates */
+	for (i = 0; i < 2; i++) {
+		const char *esteid_cert_names[2] = {"Isikutuvastus", "Allkirjastamine"};
+		const char *esteid_cert_paths[2] = {"3f00:adf1:3401", "3f00:adf2:341f"};
+		const u8 esteid_cert_ids[2] = {1, 2};
+
+		struct sc_pkcs15_cert_info cert_info;
+		struct sc_pkcs15_object cert_obj;
+
+		memset(&cert_info, 0, sizeof(cert_info));
+		memset(&cert_obj, 0, sizeof(cert_obj));
+
+		strlcpy(cert_obj.label, esteid_cert_names[i], sizeof(cert_obj.label));
+		sc_format_path(esteid_cert_paths[i], &cert_info.path);
+		cert_info.id.value[0] = esteid_cert_ids[i];
+		cert_info.id.len = 1;
+		r = sc_pkcs15emu_add_x509_cert(p15card, &cert_obj, &cert_info);
+		if (r < 0)
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+		// Read data from first cert
+		if (i != 0)
+			continue;
+
+		sc_pkcs15_cert_t *cert = NULL;
+		r = sc_pkcs15_read_certificate(p15card, &cert_info, &cert);
+		LOG_TEST_RET(card->ctx, r, "Could not read authentication certificate");
+
+		if (cert->key->algorithm == SC_ALGORITHM_EC)
+			field_length = cert->key->u.ec.params.field_length;
+
+		static const struct sc_object_id cn_oid = {{2, 5, 4, 3, -1}};
+		u8 *cn_name = NULL;
+		size_t cn_len = 0;
+		sc_pkcs15_get_name_from_dn(card->ctx, cert->subject, cert->subject_len, &cn_oid, &cn_name, &cn_len);
+		if (cn_len > 0) {
+			char *token_name = malloc(cn_len + 1);
+			if (token_name) {
+				memcpy(token_name, cn_name, cn_len);
+				token_name[cn_len] = '\0';
+				set_string(&p15card->tokeninfo->label, (const char *)token_name);
+				free(token_name);
+			}
+		}
+		free(cn_name);
+		sc_pkcs15_free_certificate(cert);
+	}
+
+	/* add pins */
+	for (i = 0; i < 3; i++) {
+		const char *esteid_pin_names[3] = {"PIN1", "PIN2", "PUK"};
```

Comment:
aren't there better names for PIN1 and PIN2?

Reply:
This is how they are labeled in PIN envelope when they are handed to citizens

Reply:
Then the *real* fix is needed elsewhere...